Small DSP accelerator fixes

leoetlino opened the pull request:
A collection of small DSP accelerator fixes. Each commit is self contained and changes are explained in their respective commit messages.

JMC47 commented:
@delroth you may want to review this one

JMC47 commented:
This break's Disney's Magical Park in HLE only.

delroth commented on Source/UnitTests/Core/DSP/DSPAcceleratorTest.cpp:
```diff
@@ -0,0 +1,159 @@
+                              [this] { accov_raised = true; }, StubReadMemory);
+}
+
+TEST(DSPAccelerator, Initialisation)
```
I think in general we use US spelling, not UK spelling. So Initialization :p

delroth commented on Source/UnitTests/Core/DSP/DSPAcceleratorTest.cpp:
```diff
@@ -0,0 +1,159 @@
+// Copyright 2017 Dolphin Emulator Project
```
<3<3<3<3<3

delroth commented on Source/Core/Core/DSP/DSPAccelerator.cpp:
```diff
@@ -72,9 +72,15 @@ void dsp_write_aram_d3(u16 value)
 }
 
 u16 ReadAccelerator(u32 start_address, u32 end_address, u32* current_address, u16 sample_format,
```
For example, if accelerator state was mirrored in there, you could implement the YN2 behavior in a cleaner way I think.

leoetlino commented:
Converted the accelerator code into a C++ class (as mentioned on IRC), and rebased all other patches.

Unfortunately, I'm still unable to figure out the regression in the Disney game... It's kind of weird that it breaks only with AX HLE... So for now, the WIP tag should be kept.

leoetlino commented:
Fixed the regression. The Windows build failure should be unrelated; it's trying to rebuild DSPTool (because I edited DSP header files), but that fails when linking for some reason :/

@delroth would you mind giving this another review?

skidau commented:
To fix the Windows build issue, add stubs to DSPHost.* and possibly remove the #include from DSPAccelerator.cpp.  Take a look at 6a6c086b502290de80f51e07c282070b5122f524 to base the changes.

lioncash commented on Source/Core/Core/DSP/DSPCore.cpp:
```diff
@@ -111,11 +113,21 @@ static void DSPCore_FreeMemoryPages()
   g_dsp.irom = g_dsp.iram = g_dsp.dram = g_dsp.coef = nullptr;
 }
 
+class LLEAccelerator : public Accelerator
```
This and the HLE accelerator can likely be made final

lioncash commented on Source/Core/Core/DSP/DSPAccelerator.h:
```diff
@@ -4,18 +4,54 @@
-u16 ReadAccelerator(u32 start_address, u32 end_address, u32* current_address, u16 sample_format,
-                    s16* yn1, s16* yn2, u16* pred_scale, s16* coefs,
-                    std::function<void()> end_exception);
+class Accelerator
```
This should probably have a virtual destructor

leoetlino replied:
Thanks, fixed both issues.

JMC47 commented:
All of the cases appear to be working now and Disney Magical Park isn't broken any more.

I'm sad it isn't broken anymore because it was amazing, but, I guess it's better not to have known regressions when merging things.

delroth commented on Source/Core/Core/DSP/DSPHWInterface.cpp:
```diff
@@ -98,7 +98,7 @@ u16 gdsp_mbox_read_l(Mailbox mbx)
 }
 
-void gdsp_ifx_write(u32 addr, u32 val)
+void gdsp_ifx_write(u32 addr, u16 val)
```
Since some compilers are broken (*cough* clang *cough*) this needs an appropriate change in our JIT to clear the upper 16 bits of the value when the function is called.

delroth commented:
LGTM. If LLE works with that JIT change, feel free to merge.
